consensus/src/consensus/mod.rs

```diff
@@ -610,6 +615,97 @@ impl ConsensusApi for Consensus {
+                            // We found the chain block we need
+                            matching_chain_block_hash = Some(hash);
+                            break;
+                        }
+                        cmp::Ordering::Greater => {
+                            high_index = mid - 1;
+                        }
+                        cmp::Ordering::Less => {
+                            low_index = mid + 1;
+                        }
+                    }
+                } else {
+                    trace!("Did not find a compact header with hash {}", hash);
+                    return None;
+                }
+            } else {
+                trace!("Did not find a hash at index {}", mid);
+                return None;
+            }
+        }
+
+        let matching_chain_block_hash = matching_chain_block_hash?;
+
+        if let Ok(acceptance_data) = self.acceptance_data_store.get(matching_chain_block_hash) {
+            let maybe_index_and_containing_acceptance =
+                acceptance_data.iter().find_map(|mbad| {
+                    let tx_arr_index = mbad.accepted_transactions.iter().enumerate().find_map(|(index, tx)| {
+                        if tx.transaction_id == txid {
+                            Some(index)
+                        } else {
+                            None
+                        }
+                    });
+
+                    tx_arr_index.map(|index| (index, mbad.clone()))
+                });
+
+            if let Some((index, containing_acceptance)) = maybe_index_and_containing_acceptance {
+                // Found Merged block containing the TXID
+                let tx = &self.block_transactions_store.get(containing_acceptance.block_hash).unwrap()[index];
+
+                if tx.id() != txid {
+                    // Should never happen, but do a sanity check. This would mean something went wrong with storing block transactions
+                    // Sanity check is necessary to guarantee that this function will never give back a wrong address (err on the side of None)
+                    warn!(
+                        "Expected {} to match {} when checking block_transaction_store using array index of transaction",
+                        tx.id(),
+                        txid
+                    );
+                    return None;
+                }
+
+                if tx.inputs.is_empty() {
+                    // A transaction may have no inputs (like a coinbase transaction)
+                    return None;
+                }
+
+                let first_input_prev_outpoint = &tx.inputs[0].previous_outpoint;
+                // Expected to never fail, since
+                let utxo_diff = self.utxo_diffs_store.get(matching_chain_block_hash).unwrap();
+                let removed_diffs = utxo_diff.removed();
+
+                return Some(removed_diffs.get(first_input_prev_outpoint)?.script_public_key.clone());
+            }
+        };
+
+        None
```

Comment:
```suggestion
 let matching_chain_block_hash = loop {
            let mid = low_index + (high_index - low_index) / 2;
            let hash = sc_read.get_by_index(mid).inspect_err(|_err| trace!("Did not find a hash at index {}", mid)).ok()?;
            let compact_header = self
                .headers_store
                .get_compact_header_data(hash)
                .inspect_err(|_err| trace!("Did not find a hash at index {}", mid))
                .ok()?;
            match compact_header.daa_score.cmp(&target_daa_score) {
                cmp::Ordering::Equal => {
                    // We found the chain block we need
                    break Some(hash);
                }
                cmp::Ordering::Greater => {
                    high_index = mid - 1;
                }
                cmp::Ordering::Less => {
                    low_index = mid + 1;
                }
            }
            if low_index > high_index {
                break None;
            }
        }?;

        let acceptance_data = self.acceptance_data_store.get(matching_chain_block_hash).ok()?;
        let (index, containing_acceptance) = acceptance_data.iter().find_map(|mbad| {
            let tx_arr_index =
                mbad.accepted_transactions.iter().enumerate().find_map(|(index, tx)| (tx.transaction_id == txid).then_some(index));
            tx_arr_index.map(|index| (index, mbad.clone()))
        })?;

        // Found Merged block containing the TXID
        let tx = &self.block_transactions_store.get(containing_acceptance.block_hash).unwrap()[index];

        if tx.id() != txid {
            // Should never happen, but do a sanity check. This would mean something went wrong with storing block transactions
            // Sanity check is necessary to guarantee that this function will never give back a wrong address (err on the side of None)
            warn!("Expected {} to match {} when checking block_transaction_store using array index of transaction", tx.id(), txid);
            return None;
        }

        if tx.inputs.is_empty() {
            // A transaction may have no inputs (like a coinbase transaction)
            return None;
        }

        let first_input_prev_outpoint = &tx.inputs[0].previous_outpoint;
        // Expected to never fail, since
        let utxo_diff = self.utxo_diffs_store.get(matching_chain_block_hash).unwrap();
        let removed_diffs = utxo_diff.removed();

        Some(removed_diffs.get(first_input_prev_outpoint)?.script_public_key.clone())
```

Reply:
Added in [8dc0874](https://github.com/kaspanet/rusty-kaspa/pull/436/commits/8dc0874523ce94707c2c294819f63291ac3dc9f4)

I couldn't apply as-is since `inspect_err` was unstable. So I replaced that with `map_err` but kept the rest.

Also added more comments.